pypeit/reduce.py

```diff
@@ -449,6 +425,36 @@ def run_extraction(self, global_sky, sobjs_obj, skymask, ra=None, dec=None, obst
         msgs.info("Generating wavelength image")
         self.waveimg = self.wv_calib.build_waveimg(self.tilts, self.slits, spat_flexure=self.spat_flexure_shift)
 
+    def run_extraction(self, global_sky, sobjs_obj, skymask, ra=None, dec=None, obstime=None, return_negative=False):
+        """
+        Primary code flow for PypeIt reductions
+
+        *NOT* used by COADD2D
+
+        Args:
+            global_sky (`numpy.ndarray`_):
+                Initial global sky model
+            sobjs_obj (:class:`pypeit.specobjs.SpecObjs`):
+                List of objects found during `run_objfind`
+            skymask (`numpy.ndarray`_):
+               Boolean image indicating which pixels are useful for global sky subtraction
+            ra (float, optional):
+                Required if helio-centric correction is to be applied
+            dec (float, optional):
+                Required if helio-centric correction is to be applied
+            obstime (:obj:`astropy.time.Time`, optional):
+                Required if helio-centric correction is to be applied
+
+        Returns:
+            tuple: skymodel (ndarray), objmodel (ndarray), ivarmodel (ndarray),
+               outmask (ndarray), sobjs (SpecObjs), waveimg (`numpy.ndarray`_),
+               tilts (`numpy.ndarray`_).
+               See main doc string for description
+
+        """
+        # Start by preparing some masks and the wavelength image, ready for extraction
```

Comment:
Hi @rcooke-ast, 

Thanks for trying to address this. My main object is that the run_objfind is doing some of the same things as prepare_extraction, i.e. since it sky-subtracts it needs to have the tilts as well. So we are generating the tilts and dealing with the flexure etc. twice. This is not really your fault, i.e. the code was refactored recently to deal with DEIMOS subtleties and ended up in this form. Reduce is doing way to many things and the control flow is getting more and more opaque as we try to tailor the steps to some required applications (i.e. no extraction, no local, etc.) I would like to refactor reduce altogether into an object finding class and a separate extraction class. That is on my to do list. For now, I'm fine with your changes as long as they pass tests. If you want to think about how to reduce some of the redundancy in run_objfind and run_extraction you can but I'll approve this as is.